source/MaterialXContrib/MaterialXNode/MaterialXTextureOverride.cpp

```diff
@@ -36,24 +37,78 @@ MaterialXTextureOverride::MaterialXTextureOverride(const MObject& obj)
+        MaterialX::ElementPtr element = _document->getDescendant(_element.asChar());
+        MaterialX::ShaderRefPtr shaderRef = element->asA<MaterialX::ShaderRef>();
+        MaterialX::OutputPtr output = element->asA<MaterialX::Output>();
+        if (!output && !shaderRef)
+        {
+            // Should never occur as we pre-filter renderables before creating the node + override
+            throw MaterialX::Exception("MaterialXTextureOverride: Invalid type to create wrapper for");
+        }
 
-			std::cout << "MaterialXTextureOverride: Add XML fragment to manager: " << materialXData.fragmentName << std::endl;
-		}
-		catch (MaterialX::Exception& e)
-		{
-			std::cerr << "MaterialXTextureOverride: Failed to generate XML wrapper: " << e.what() << std::endl;
-		}
+        std::vector<MaterialX::GenContext*> contexts;
+        MaterialX::GenContext* glslContext = new MaterialX::GenContext(MaterialX::GlslShaderGenerator::create());
 
-		// TODO: Use our own image loader vs Mayas.
-		// MaterialX::StbImageLoaderPtr stbLoader = MaterialX::StbImageLoader::create();
+        // Stop emission of environment map lookups.
+        glslContext->registerSourceCodeSearchPath(libSearchPath);
+        if (shaderRef)
+        {
+            glslContext->getOptions().hwSpecularEnvironmentMethod = MaterialX::SPECULAR_ENVIRONMENT_FIS;
+            glslContext->getOptions().hwMaxActiveLightSources = 0;
+        }
+        else
+        {
+            glslContext->getOptions().hwSpecularEnvironmentMethod = MaterialX::SPECULAR_ENVIRONMENT_NONE;
+        }
+        glslContext->getOptions().fileTextureVerticalFlip = true;
+
+        _glslWrapper = new MaterialX::OGSXMLFragmentWrapper(glslContext);
+        _glslWrapper->setOutputVertexShader(true);     
+
+        // TODO: This just indicates that lighting is required. As direct lighting
+        // is not supported, the requirement means that indirect lighting is required.
+        // bool requiresLighting = (shaderRef != nullptr);
+        std::cout << "MaterialXTextureOverride: Create XML wrapper" << std::endl;
+        _glslWrapper->createWrapper(element);
+        // Get the fragment name
+        _fragmentName.set(_glslWrapper->getFragmentName().c_str());
+
+        // Register fragments with the manager if needed
+        //
+        MHWRender::MRenderer* theRenderer = MHWRender::MRenderer::theRenderer();
+        if (theRenderer)
+        {
+            MHWRender::MFragmentManager* fragmentMgr =
+                theRenderer->getFragmentManager();
+            if (fragmentMgr)
+            {
+                bool fragmentExists = (_fragmentName.length() > 0) && fragmentMgr->hasFragment(_fragmentName);
+                if (!fragmentExists)
+                {
+                    std::stringstream glslStream;
+                    _glslWrapper->getDocument(glslStream);
+                    std::string xmlFileName(Plugin::instance().getResourcePath().asString() + "/standard_surface_default.xml");
```

Comment:
There's lots of changes in this file so could you please summarize them a bit? It looks like this version still contains a surface node experiment? Isn't it time to move that to MaterialXSurfaceOveride now that it's there? Or are there some valuable texture-specific changes as well?

Reply:
This is the easiest way to do the merge  for now. Also the code can be used for textures as well as for shaders so it will need to be made common.